espnet2/enh/layers/tcndenseunet.py

```diff
@@ -0,0 +1,382 @@
+        else:
+            conv = torch.nn.Conv2d(
+                in_channels, out_channels, ksz, stride, padding, padding_mode="reflect"
+            )
+        act = activations.get(activation)()
+        norm = torch.nn.GroupNorm(out_channels, out_channels, eps=1e-8)
+        self.layer = torch.nn.Sequential(conv, act, norm)
+
+    def forward(self, inp):
+        return self.layer(inp)
+
+
+class FreqWiseBlock(torch.nn.Module):
+    def __init__(self, in_channels, num_freqs, out_channels, activation=torch.nn.ELU):
+        super(FreqWiseBlock, self).__init__()
+
+        self.bottleneck = Conv2DActNorm(
+            in_channels, out_channels, (1, 1), (1, 1), (0, 0), activation=activation
+        )
+        self.freq_proc = Conv2DActNorm(
+            num_freqs, num_freqs, (1, 1), (1, 1), (0, 0), activation=activation
+        )
+
+    def forward(self, inp):
+        # bsz, chans, x, y
+
+        out = self.freq_proc(self.bottleneck(inp).permute(0, 3, 2, 1)).permute(
+            0, 3, 2, 1
+        )
+
+        return out
+
+
+class DenseBlock(torch.nn.Module):
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        num_freqs,
+        pre_blocks=2,
+        freq_proc_blocks=1,
+        post_blocks=2,
+        ksz=(3, 3),
+        activation=torch.nn.ELU,
+        hid_chans=32,
+    ):
+
+        super(DenseBlock, self).__init__()
+
+        assert post_blocks >= 1
+        assert pre_blocks >= 1
+
+        self.pre_blocks = torch.nn.ModuleList([])
+        tot_layers = 0
+        for indx in range(pre_blocks):
+            c_layer = Conv2DActNorm(
+                in_channels + hid_chans * tot_layers,
+                hid_chans,
+                ksz,
+                (1, 1),
+                (1, 1),
+                activation=activation,
+            )
+            self.pre_blocks.append(c_layer)
+            tot_layers += 1
+
+        self.freq_proc_blocks = torch.nn.ModuleList([])
+        for indx in range(freq_proc_blocks):
+            c_layer = FreqWiseBlock(
+                in_channels + hid_chans * tot_layers,
+                num_freqs,
+                hid_chans,
+                activation=activation,
+            )
+            self.freq_proc_blocks.append(c_layer)
+            tot_layers += 1
+
+        self.post_blocks = torch.nn.ModuleList([])
+        for indx in range(post_blocks - 1):
+            c_layer = Conv2DActNorm(
+                in_channels + hid_chans * tot_layers,
+                hid_chans,
+                ksz,
+                (1, 1),
+                (1, 1),
+                activation=activation,
+            )
+            self.post_blocks.append(c_layer)
+            tot_layers += 1
+
+        last = Conv2DActNorm(
+            in_channels + hid_chans * tot_layers,
+            out_channels,
+            ksz,
+            (1, 1),
+            (1, 1),
+            activation=activation,
+        )
+        self.post_blocks.append(last)
+
+    def forward(self, input):
```

Comment:
what type of GPU are you using? this needs many memeory usage. This is just a comment, not a problem at all.

Reply:
NVIDIA A100 40GB. Is it more memory intensive than using e.g. a torch Sequential wrapper ?

Reply:
I meant the densenet concats features maps and uses many memory. There are some memory-efficient ways to reduce the memory consumption by re-computing some tensors during back-propagation.

Reply:
like this for example ? https://github.com/gpleiss/efficient_densenet_pytorch/blob/master/models/densenet.py not really familiar with densenets

Reply:
Yeah. it is what I was talking about. they used torch.utils.checkpoint to save training-time memory.